### PR TITLE
Fixes #134 : add Exception on HeadeMeta::setCharset() when the doctype is Xhtml

### LINKED ISSUE #134: Setting charset in Doctype::XHTML1_RDFA fails
I'm getting an error when I set the charset on a rdfa document. Is that expected? Or an oversight?

```
        // set doctype
        $this->helperPluginManager->get('doctype')
            ->setDoctype(\Zend\View\Helper\Doctype::XHTML1_RDFA);

$this->helperPluginManager->get('headMeta')
            ->setCharset('utf-8');
```

```
Invalid value passed to set; please use setMeta()
./vendor/zendframework/zend-view/src/Helper/HeadMeta.php:442
```

Further; following the suggestion in the error gives me an error too.

```
$this->helperPluginManager->get('headMeta')
         ->setMeta('charset', 'utf-8');
```

```
Method "setMeta" does not exist
./vendor/zendframework/zend-view/src/Helper/Placeholder/Container/AbstractStandalone.php:83
```

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes #134

### CODE REVIEW

File: src/Helper/HeadMeta.php [hunk 1]
```diff
@@ -458,10 +458,15 @@ public function set($value)
      * Not valid in a non-HTML5 doctype
      *
      * @param  string $charset
+     * @throws Exception\RuntimeException
      * @return HeadMeta Provides a fluent interface
      */
     public function setCharset($charset)
     {
+        if ($this->view->plugin('doctype')->isXhtml()) {
+            throw new Exception\RuntimeException('XHTML* doctype has no attribute charset');
```
Comment (froschdesign) [inline src/Helper/HeadMeta.php]: Please move this check to the `isValid` method and call `isValid` before `$this->set($item);` in the current method.

Comment (froschdesign) [reply src/Helper/HeadMeta.php]: You should also extend the exception message:

> 'XHTML* doctype has no attribute charset; please use appendHttpEquiv'

Then change the exception class to `Exception\InvalidArgumentException`.

**Comment by froschdesign** [comment]
@samsonasik 
The bug report includes another problem:

> `->setMeta('charset', 'utf-8');`
> > Method "setMeta" does not exist

The exception message is wrong:

https://github.com/zendframework/zend-view/blob/4d9954c05a3981ca711bd55932dc1d2b55580392/src/Helper/HeadMeta.php#L442

Remove the hint to the method "setMeta" and update the DocBlock. Add a short example and fix also the data type for the value parameter.
The same for the methods `prepend` and `append`.

Thanks in advance!

File: src/Helper/HeadMeta.php [hunk 2]
```diff
@@ -334,22 +335,27 @@ protected function isValid($item)
             return false;
         }
 
+        $doctype = $this->view->plugin('doctype');
+        if ($item->type === 'charset' && $doctype->isXhtml()) {
+            throw new Exception\RuntimeException('XHTML* doctype has no attribute charset; please use appendHttpEquiv()');
```
Comment (froschdesign) [inline src/Helper/HeadMeta.php]: `return false` here.

File: src/Helper/HeadMeta.php (hunk 1 above)
Comment (samsonasik) [reply src/Helper/HeadMeta.php]: updated

File: src/Helper/HeadMeta.php [hunk 3]
```diff
@@ -458,20 +464,17 @@ public function set($value)
...
-        if ($this->view->plugin('doctype')->isXhtml()) {
-            throw new Exception\RuntimeException('XHTML* doctype has no attribute charset');
-        }
-
         $item = new stdClass;
         $item->type = 'charset';
         $item->charset = $charset;
         $item->content = null;
         $item->modifiers = [];
+        $this->isValid($item);
```
Comment (froschdesign) [inline src/Helper/HeadMeta.php]: If false, then throw `Exception\InvalidArgumentException` here.

Comment (froschdesign) [reply src/Helper/HeadMeta.php]: Compare with the methods `set`, `prepend` etc.

**Comment by samsonasik** [comment]
I've updated exception messages and docblock param in HeadMeta::set(), append(), prepend(), do you have suggestion for data type parameter ? the stdClass check already in `isValid()` method so I think the type parameter should not be used, instead only update the docblock parameter.

File: src/Helper/HeadMeta.php (hunk 2 above)
Comment (samsonasik) [reply src/Helper/HeadMeta.php]: already throw an Exception, need a return false ?

Comment (samsonasik) [reply src/Helper/HeadMeta.php]: understood, make a RuntimeException in `setCharset()` ?

File: src/Helper/HeadMeta.php (hunk 3 above)
Comment (samsonasik) [reply src/Helper/HeadMeta.php]: updated

**Comment by froschdesign** [comment]
> …instead only update the docblock parameter.

That's what i meant!

File: src/Helper/HeadMeta.php [hunk 4]
```diff
@@ -474,7 +473,11 @@ public function setCharset($charset)
         $item->charset = $charset;
         $item->content = null;
         $item->modifiers = [];
-        $this->isValid($item);
+
+        if (! $this->isValid($item)) {
+            throw new Exception\RuntimeException('XHTML* doctype has no attribute charset; please use appendHttpEquiv()');
```
Comment (froschdesign) [inline src/Helper/HeadMeta.php]: You don't like the `InvalidArgumentException`?! 😉

Comment (samsonasik) [reply src/Helper/HeadMeta.php]: updated to InvalidArgumentException

**Comment by froschdesign** [comment]
@samsonasik 
Thanks for your contribution! 👍

**Pull request merged**